File: examples/random_example.cpp
/**
 * @file random_example.cpp
 * @brief Example demonstrating the usage of the random module
 */

#include <cstddef>
#include <numeric>
#include <print>
#include <string_view>

#include "containers/container.hpp"
#include "random/random_gen.hpp"

using cpp_features::containers::Container;
using cpp_features::random::RandomGenerator;
using cpp_features::random::SampleFromRange;
using cpp_features::random::ShuffleContainer;

namespace {

/**
 * @brief Demonstrate basic random number generation
 */
void DemonstrateBasicGeneration() {
  std::println("\n1. Basic Random Generation:");
  RandomGenerator generator{42};  // Fixed seed for reproducible output

  // Integer generation
  auto dice_roll = generator.GenerateInt(1, 6);
  auto lottery_number = generator.GenerateInt(1, 49);
  std::println("   Dice roll: {}, Lottery number: {}", dice_roll, lottery_number);

  // Float generation
  auto probability = generator.GenerateReal(0.0, 1.0);
  auto temperature = generator.GenerateReal(-10.0F, 40.0F);
  std::println("   Probability: {:.3f}, Temperature: {:.1f}°C", probability, temperature);

  // Boolean generation
  std::string_view coin_flip{generator.GenerateBool() ? "Heads" : "Tails"};
  std::string_view rare_event{generator.GenerateBool(0.1) ? "Occurred" : "Didn't occur"};
  std::println("   Coin flip: {}, Rare event: {}", coin_flip, rare_event);
}

/**
 * @brief Demonstrate various probability distributions
 */
void DemonstrateProbabilityDistributions() {
  std::println("\n2. Probability Distributions:");
  RandomGenerator generator{42};

  // Uniform distributions
  Container uniform_ints(generator.GenerateIntVector(10, 20, 5));
  std::println("   Uniform integers [10, 20]: {}", uniform_ints);

  Container uniform_reals(generator.GenerateRealVector(0.5F, 1.5F, 3));
  std::println("   Uniform reals [0.5, 1.5): {}", uniform_reals);

  // Normal distribution
  Container<int> iq_scores;
  for (std::size_t i = 0; i < 5; ++i) {
    iq_scores.Add(static_cast<int>(generator.GenerateNormal(100.0, 15.0)));
  }
  std::println("   IQ scores (μ = 100, σ = 15): {}", iq_scores);

  // Bernoulli distribution
  Container<std::string_view> coin_flips;
  for (std::size_t i = 0; i < 10; ++i) {
    coin_flips.Add(generator.GenerateBool(0.7) ? "H" : "T");
  }
  std::println("   Biased coin flips (p = 0.7): {}", coin_flips);
}

/**
 * @brief Demonstrate seeding and reproducibility
 */
void DemonstrateSeedingAndReproducibility() {
  std::println("\n3. Seeding and Reproducibility:");
  RandomGenerator gen1{123};
  RandomGenerator gen2{123};
  auto val1 = gen1.GenerateInt(1, 1000);
  auto val2 = gen2.GenerateInt(1, 1000);
  std::println("   Same seed (123): {} == {}", val1, val2);

  RandomGenerator gen3;  // Random seed
  gen3.SeedWithTime();
  auto time_seeded = gen3.GenerateInt(1, 1000);
  std::println("   Time-seeded value: {}", time_seeded);
}

/**
 * @brief Demonstrate range operations (shuffling and sampling)
 */
void DemonstrateRangeOperations() {
  std::println("\n4. Range Operations:");

  // Container shuffling
  Container deck{1, 2, 3, 4, 5, 6, 7, 8};
  std::println("   Original deck: {}", deck);

  ShuffleContainer(deck);
  std::println("   Shuffled deck: {}", deck);

  // Random sampling
  Container population{10, 20, 30, 40, 50, 60, 70, 80, 90};
  Container sample(SampleFromRange(population, 4));
  std::println("   Sample from population: {}", sample);

  // String sampling
  std::string_view vowels{"AEIOU"};
  Container vowel_sample(SampleFromRange(vowels, 3));
  std::println("   Random vowels: {}", vowel_sample);
}

/**
 * @brief Demonstrate practical applications of random generation
 */
void DemonstratePracticalApplications() {
  std::println("\n5. Practical Applications:");
  RandomGenerator dice_generator;

  // Simulate rolling multiple dice
  Container dice_rolls(dice_generator.GenerateIntVector(1, 6, 3));
  auto total = std::reduce(dice_rolls.begin(), dice_rolls.end(), 0);
  std::println("   Rolling 3 dice: {}, Total: {}", dice_rolls, total);

  // Monte Carlo estimation (simple π approximation)
  auto inside_circle = 0;
  constexpr auto SAMPLES = 10000;
  for (std::size_t i = 0; i < SAMPLES; ++i) {
    auto x = dice_generator.GenerateReal(-1.0, 1.0);
    auto y = dice_generator.GenerateReal(-1.0, 1.0);
    if (x * x + y * y <= 1.0) {
      ++inside_circle;
    }
  }
  auto pi_estimate = 4.0 * inside_circle / SAMPLES;
  std::println("   Monte Carlo π estimation ({} samples): {:.4f}", SAMPLES, pi_estimate);
}

}  // namespace

auto main() -> int {
  std::println("=== Random Module Example ===");

  DemonstrateBasicGeneration();
  DemonstrateProbabilityDistributions();
  DemonstrateSeedingAndReproducibility();
  DemonstrateRangeOperations();
  DemonstratePracticalApplications();

  std::println("\n=== Random Module Example Completed ===");
  return 0;
}

File: python/examples/random_example.py
"""Example demonstrating the usage of the random module."""

from demo.containers import Container
from demo.random import RandomGenerator, sample, shuffle


def demonstrate_basic_generation() -> None:
    """Demonstrate basic random number generation."""
    print('\n1. Basic Random Generation:')
    generator = RandomGenerator(seed=42)  # Fixed seed for reproducible output

    # Integer generation
    dice_roll = generator.rand_int(1, 6)
    lottery_number = generator.rand_int(1, 49)
    print(f'   Dice roll: {dice_roll}, Lottery number: {lottery_number}')

    # Float generation
    probability = generator.rand_float(0.0, 1.0)
    temperature = generator.rand_float(-10.0, 40.0)
    print(f'   Probability: {probability:.3f}, Temperature: {temperature:.1f}°C')

    # Boolean generation
    coin_flip = 'Heads' if generator.rand_bool() else 'Tails'
    rare_event = 'Occurred' if generator.rand_bool(0.1) else "Didn't occur"
    print(f'   Coin flip: {coin_flip}, Rare event: {rare_event}')


def demonstrate_probability_distributions() -> None:
    """Demonstrate various probability distributions."""
    print('\n2. Probability Distributions:')
    generator = RandomGenerator(seed=42)

    # Uniform distributions
    uniform_ints = generator.rand_ints(10, 20, count=5)
    print(f'   Uniform integers [10, 20]: {uniform_ints}')

    uniform_reals = generator.rand_floats(0.5, 1.5, count=3)
    print(f'   Uniform reals [0.5, 1.5): {uniform_reals}')

    # Normal distribution
    iq_scores: list[int] = []
    for _ in range(5):
        iq_scores.append(int(generator.normal(100.0, 15.0)))
    print(f'   IQ scores (μ = 100, σ = 15): {iq_scores}')

    # Bernoulli distribution
    coin_flips: list[str] = []
    for _ in range(10):
        coin_flips.append('H' if generator.rand_bool(0.7) else 'T')
    print(f'   Biased coin flips (p = 0.7): {coin_flips}')


def demonstrate_seeding_and_reproducibility() -> None:
    """Demonstrate seeding and reproducibility."""
    print('\n3. Seeding and Reproducibility:')
    rg1 = RandomGenerator(seed=123)
    value1 = rg1.rand_int(1, 1000)
    rg2 = RandomGenerator(seed=123)
    value2 = rg2.rand_int(1, 1000)
    print(f'   Same seed (123): {value1} == {value2}')

    rg3 = RandomGenerator()  # Random seed
    rg3.seed_with_time()
    time_seeded = rg3.rand_int(1, 1000)
    print(f'   Time-seeded value: {time_seeded}')


def demonstrate_range_operations() -> None:
    """Demonstrate range operations (shuffling and sampling)."""
    print('\n4. Range Operations:')

    # Container shuffling
    deck = Container(int, [x for x in range(1, 9)])
    print(f'   Original deck: {deck}')

    shuffled_deck = Container(int, [x for x in range(1, 9)])
    shuffle(shuffled_deck)
    print(f'   Shuffled deck: {shuffled_deck}')

    # Random sampling
    population = Container(int, [x for x in range(10, 100, 10)])
    sample_result = Container(int, sample(population, 4))
    print(f'   Sample from population: {sample_result}')

    # String sampling
    vowels = Container(str, ['A', 'E', 'I', 'O', 'U'])
    vowel_sample = Container(str, sample(vowels, 3))
    print(f'   Random vowels: {vowel_sample}')


def demonstrate_practical_applications() -> None:
    """Demonstrate practical applications of random generation."""
    print('\n5. Practical Applications:')
    dice_generator = RandomGenerator()

    # Simulate rolling multiple dice
    dice_rolls = Container(int, dice_generator.rand_ints(1, 6, count=3))
    total = sum(dice_rolls)
    print(f'   Rolling 3 dice: {dice_rolls}, Total: {total}')

    # Monte Carlo estimation (simple π approximation)
    inside_circle = 0
    samples = 10000
    for _ in range(samples):
        x = dice_generator.rand_float(-1.0, 1.0)
        y = dice_generator.rand_float(-1.0, 1.0)
        if x * x + y * y <= 1.0:
            inside_circle += 1

    pi_estimate = 4.0 * inside_circle / samples
    print(f'   Monte Carlo π estimation ({samples} samples): {pi_estimate:.3f}')


def main() -> None:
    """Run all random examples."""
    print('=== Random Module Example ===')

    demonstrate_basic_generation()
    demonstrate_probability_distributions()
    demonstrate_seeding_and_reproducibility()
    demonstrate_range_operations()
    demonstrate_practical_applications()

    print('\n=== Random Module Example Completed ===')


if __name__ == '__main__':
    main()
